letsencrypt/auth_handler.py

```diff
@@ -66,19 +66,31 @@ def get_authorizations(self, domains, best_effort=False):
 
         self._choose_challenges(domains)
 
+        failed_domains = set()
+
         # While there are still challenges remaining...
         while self.achalls:
             resp = self._solve_challenges()
             logger.info("Waiting for verification...")
 
-            # Send all Responses - this modifies achalls
-            self._respond(resp, best_effort)
+            # Send all Responses - this modifies dv_c and cont_c
+            response = self._respond(resp, best_effort)
+
+            if response:
+                failed_domains = failed_domains.union(response)
+                for domain in failed_domains:
+                    logger.warning(
+                        "Challenge failed for domain %s",
+                        domain)
+
+        returnDomains = [domain for domain in domains
+                        if domain not in failed_domains]
 
         # Just make sure all decisions are complete.
         self.verify_authzr_complete()
         # Only return valid authorizations
         return [authzr for authzr in self.authzr.values()
-                if authzr.body.status == messages.STATUS_VALID]
+                if authzr.body.status == messages.STATUS_VALID], returnDomains
```

Comment:
We should check that authorizations succeeded for at least one of the requested names. Otherwise, we get errors like this later:
```
AssertionError: Must provide one or more hostnames for the CSR.
```
To do this, you can store the result of the list comprehension in a local variable and raise an exception if it is empty.